packages/itwin/tree-widget/src/components/trees/models-tree/HierarchyBasedVisibilityHandler.ts

```diff
@@ -0,0 +1,848 @@
+  hierarchyAutoUpdateEnabled?: boolean;
+}
+
+/**
+ * Hierarchy based visibility handler.
+ * When determining visibility for nodes, it should take into account the visibility of their children.
+ * @public
+ */
+export interface IHierarchyBasedVisibilityHandler extends IVisibilityHandler {
+  filteredDataProvider?: IFilteredPresentationTreeDataProvider;
+}
+
+/**
+ * Creates an instance if [[IHierarchyBasedVisibilityHandler]].
+ * @public
+ */
+export function createHierarchyBasedVisibilityHandler(props: HierarchyBasedVisibilityHandlerProps): IHierarchyBasedVisibilityHandler {
+  return new VisibilityHandlerImplementation(props);
+}
+
+const MAX_PARALLEL_SUBJECT_MODELS_REQUESTS = 1;
+const MAX_PARALLEL_MODELS_CATEGORIES_REQUESTS = 1;
+
+class VisibilityHandlerImplementation implements IVisibilityHandler {
+  public filteredDataProvider?: IFilteredPresentationTreeDataProvider;
+  public readonly isHierarchyBased = true;
+  private readonly _eventListener: IVisibilityChangeEventListener;
+  private readonly _queryHandler: IQueryHandler;
+  private _removePresentationHierarchyListener?: () => void;
+
+  constructor(private readonly _props: HierarchyBasedVisibilityHandlerProps) {
+    this._eventListener = createVisibilityChangeEventListener(_props.viewport);
+    this._queryHandler = createQueryHandler(this._props.viewport.iModel);
+    // istanbul ignore if
+    if (this._props.hierarchyAutoUpdateEnabled) {
+      // eslint-disable-next-line @itwin/no-internal
+      this._removePresentationHierarchyListener = Presentation.presentation.onIModelHierarchyChanged.addListener(() => {
+        this._queryHandler.invalidateCache();
+      });
+    }
+  }
+
+  // istanbul ignore next
+  public get onVisibilityChange() {
+    return this._eventListener.onVisibilityChange;
+  }
+
+  public async getVisibilityStatus(node: TreeNodeItem): Promise<VisibilityStatus> {
+    return firstValueFrom(this.getVisibilityStatusObs(node));
+  }
+
+  public async changeVisibility(node: TreeNodeItem, shouldDisplay: boolean): Promise<void> {
+    return toVoidPromise(this.changeVisibilityObs(node, shouldDisplay));
+  }
+
+  // istanbul ignore next
+  public dispose(): void {
+    this._eventListener.dispose();
+    this._removePresentationHierarchyListener?.();
+  }
+
+  private getVisibilityStatusObs(node: TreeNodeItem): Observable<VisibilityStatus> {
+    if (!isPresentationTreeNodeItem(node)) {
+      return createVisibilityStatusObs("disabled");
+    }
+
+    const nodeKey = node.key;
+    if (NodeKey.isClassGroupingNodeKey(nodeKey)) {
+      return from(this.getElementGroupingNodeDisplayStatus(nodeKey));
+    }
+
+    if (!NodeKey.isInstancesNodeKey(nodeKey)) {
+      return createVisibilityStatusObs("disabled");
+    }
+
+    if (NodeUtils.isSubjectNode(node)) {
+      // note: subject nodes may be merged to represent multiple subject instances
+      return this.getSubjectNodeVisibilityStatus(node, from(nodeKey.instanceKeys).pipe(map((key) => key.id)));
```

Comment:
We'd still have to convert this to an observable in `getSubjectNodeVisibilityStatus`, because `subjectIds` are piped to `queryhandler.querySubjectModels`. Why not do it here and utilize the `map` operator?

Reply:
You could utility the `Array.map` :) But okay, this is not a big deal.